### ISSUE TITLE
Bleadperl breaks Test::Exception 0.27_04

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#9942]
Migrated from [rt.perl.org#70161](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=70161) (status was 'rejected')

Searchable as RT70161$

**Comment by p5pRT** [comment]
### From @andk

#### Created by sand@k81.nonet


Commit GitLive\-blead\-1583\-gba7a454 breaks the test suite of
Test&#8203;::Exception\.

Author&#8203;: Rafael Garcia\-Suarez \<rgarciasuarez@&#8203;gmail\.com>
Date&#8203;:   Tue Jul 14 09&#8203;:16&#8203;:17 2009 \+0200

&nbsp;    Merge Carp&#8203;::Heavy into Carp

I've reported the issue in
https://rt.cpan.org/Ticket/Display.html?id=49617

Sorry for the delay\, apparently I forgot to inform p5p about it at that
time\.

<details><summary>Perl Info</summary>

```

Flags:
    category=library
    severity=high
    module=Carp

Site configuration information for perl 5.11.0:

Configured by sand at Fri Sep 11 08:12:58 CEST 2009.

Summary of my perl5 (revision 5 version 11 subversion 0) configuration:
  Commit id: ba7a45496628db970f8607e67b4345e8bc45f6fe
  Platform:
    osname=linux, osvers=2.6.30-1-amd64, archname=x86_64-linux-ld
    uname='linux k81 2.6.30-1-amd64 #1 smp sat aug 15 18:09:19 utc 2009 x86_64 gnulinux '
    config_args='-Dprefix=/home/src/perl/repoperls/installed-perls/perl/GitLive-blead-1583-gba7a454 -Dinstallusrbinperl=n -Uversiononly -Dusedevel -des -Ui_db -Duselongdouble'
    hint=recommended, useposix=true, d_sigaction=define
    useithreads=undef, usemultiplicity=undef
    useperlio=define, d_sfio=undef, uselargefiles=define, usesocks=undef
    use64bitint=define, use64bitall=define, uselongdouble=define
    usemymalloc=n, bincompat5005=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -pipe -fstack-protector -I/usr/local/include'
    ccversion='', gccversion='4.3.4', gccosandvers=''
    intsize=4, longsize=8, ptrsize=8, doublesize=8, byteorder=12345678
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=8, nvtype='long double', nvsize=16, Off_t='off_t', lseeksize=8
    alignbytes=16, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -fstack-protector -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib /lib64 /usr/lib64
    libs=-lnsl -lgdbm -ldb -ldl -lm -lcrypt -lutil -lc -lgdbm_compat
    perllibs=-lnsl -ldl -lm -lcrypt -lutil -lc
    libc=/lib/libc-2.9.so, so=so, useshrplib=false, libperl=libperl.a
    gnulibc_version='2.9'
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-Wl,-E'
    cccdlflags='-fPIC', lddlflags='-shared -O2 -L/usr/local/lib -fstack-protector'

Locally applied patches:
    PERL_GIT_UNPUSHED_COMMITS    	/* do not remove this line */
    PERL_GIT_UNCOMMITTED_CHANGES	/* do not remove this line */


@INC for perl 5.11.0:
    /home/src/perl/repoperls/installed-perls/perl/GitLive-blead-1583-gba7a454/lib/site_perl/5.11.0/x86_64-linux-ld
    /home/src/perl/repoperls/installed-perls/perl/GitLive-blead-1583-gba7a454/lib/site_perl/5.11.0
    /home/src/perl/repoperls/installed-perls/perl/GitLive-blead-1583-gba7a454/lib/5.11.0/x86_64-linux-ld
    /home/src/perl/repoperls/installed-perls/perl/GitLive-blead-1583-gba7a454/lib/5.11.0
    .


Environment for perl 5.11.0:
    HOME=/home/sand
    LANG=en_US.UTF-8
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home/sand/bin:/usr/local/bin:/usr/bin:/bin:/usr/bin/X11:/usr/games:/usr/local/perl/bin:/usr/X11/bin:/sbin:/usr/sbin
    PERL_BADLANG (unset)
    SHELL=/usr/bin/zsh

-- 
andreas

```
</details>

**Comment by p5pRT** [comment]
### From @andk

>>>>> On Sun\, 01 Nov 2009 18&#8203;:06&#8203;:03 \-0800\, \(Andreas J\. Koenig\) \(via RT\) \<perlbug\-followup@&#8203;perl\.org> said&#8203;:

&nbsp;  > Commit GitLive\-blead\-1583\-gba7a454 breaks the test suite of
&nbsp;  > Test&#8203;::Exception\.

Another victim is Test&#8203;::Warn 0\.21\. The number of dependencies on these
two modules is huge\, CPANTS says >600\, http&#8203;://ali\.as/top100/ volatile
100 list scores them together with >3300 points\.

I'm not sure why Carp&#8203;::Heavy should go\, maybe it is not worth breaking
such a large portion of CPAN?

\-\- 
andreas

**Comment by p5pRT** [comment]
### From @nwc10

On Tue\, Nov 03\, 2009 at 08&#8203;:34&#8203;:37AM \+0100\, Andreas J\. Koenig wrote&#8203;:
> >>>>> On Sun\, 01 Nov 2009 18&#8203;:06&#8203;:03 \-0800\, \(Andreas J\. Koenig\) \(via RT\) \<perlbug\-followup@&#8203;perl\.org> said&#8203;:
> 
>   > Commit GitLive\-blead\-1583\-gba7a454 breaks the test suite of
>   > Test&#8203;::Exception\.
> 
> Another victim is Test&#8203;::Warn 0\.21\. The number of dependencies on these
> two modules is huge\, CPANTS says >600\, http&#8203;://ali\.as/top100/ volatile
> 100 list scores them together with >3300 points\.
> 
> I'm not sure why Carp&#8203;::Heavy should go\, maybe it is not worth breaking
> such a large portion of CPAN?

The thread that led to it is here&#8203;:

http&#8203;://www\.xray\.mpe\.mpg\.de/mailing\-lists/perl5\-porters/2009\-06/msg00978\.html

The bug report \(\#42329\) is older\.

The gist of the bug is that the old structure of Carp/Carp&#8203;::Heavy meant that
when an error occurred\, Carp would need to use \*more\* resources to report the
error\, and if the resource allocation failed\, then Carp would fail to report
the error\. This is extremely undesirable\.

So\, yes\, sadly\, it is a necessary fix\.

That leaves 2 modules on CPAN that need adapting\. Although it's unclear to
me why they are sensitive to the internals of Carp\.

Nicholas Clark

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @ap

\* Nicholas Clark \<nick@&#8203;ccl4\.org> \[2009\-11\-03 09&#8203;:45\]&#8203;:
> Although it's unclear to me why they are sensitive to the
> internals of Carp\.

That was the first question that came to my mind\.

Regards\,
\-\- 
Aristotle Pagaltzis // \<http&#8203;://plasmasturm\.org/>

**Comment by p5pRT** [comment]
### From @demerphq

2009/11/4 Aristotle Pagaltzis \<pagaltzis@&#8203;gmx\.de>&#8203;:
> \* Nicholas Clark \<nick@&#8203;ccl4\.org> \[2009\-11\-03 09&#8203;:45\]&#8203;:
>> Although it's unclear to me why they are sensitive to the
>> internals of Carp\.
>
> That was the first question that came to my mind\.

Me too\, a superficiail review shows nothing obvious at all\.

Yves


\-\- 
perl \-Mre=debug \-e "/just|another|perl|hacker/"

**Comment by p5pRT** [comment]
### From @xdg

The change moved the use of "caller\(\)" from Carp&#8203;::Heavy to Carp\.  If
Carp is loaded before Sub&#8203;::Uplevel\, caller\(\) is bound to CORE&#8203;::caller\,
not the CORE&#8203;::GLOBAL&#8203;::caller override that Sub&#8203;::Uplevel installs\.

The "fix" is to ensure that Sub&#8203;::Uplevel loads early in the test \(and to
delay loading of Carp in Test&#8203;::Exception as a general good practice\)\. 
Patch attached to the cpan RT ticket&#8203;:

https://rt.cpan.org/Ticket/Display.html?id=49617

I have also sent a patch to p5p that "fixes" Carp\.

**Comment by p5pRT** [comment]
@rgs - Status changed from 'open' to 'rejected'
